Update to golang-1.23 and also update the dependencies

praveenkumar opened the pull request.

praveenkumar committed:
go.mod: Update to go-1.23.0

crc is moved to use golang 1.23 and this project haven't been
released/rebuilt in a long time, the latest binaries were built using go 1.21,
would be good to update deps/golang version and build/release it.

go-1.22 is also EOL and dep modules shifted to use go-1.23 like `sys`
module.

- https://endoflife.date/go

praveenkumar committed:
Update all the module dependencies except libvirt

libvirt version on RHEL-8 is still libvirt-8.0.0 so we need to keep the
go module accordingly.

praveenkumar committed:
CI: Update the action/upload-artifact to v4 from v3

v3 is depericated https://github.com/actions/upload-artifact

cfergeau commented on go.mod:
```diff
@@ -1,18 +1,18 @@
-	libvirt.org/go/libvirt v1.8000.0
-	libvirt.org/go/libvirtxml v1.9008.0
+	github.com/stretchr/testify v1.9.0
+	libvirt.org/go/libvirt v1.11001.0
```
careful with libvirt upgrades, we need a version that is available in rhel8 or 9

praveenkumar replied:
I didn't update it because in RHEL-8 it is still libvirt version is 8.0

cfergeau reviewed (commented):
Have you checked if upgrading to `actions/upload-artifact/v4` in our github action workflows helps with the CI failure?
edit: just saw your update :)

praveenkumar committed:
Switch to using centos stream images from almalinux base-8

almalinux base-8 images are based on RHEL-8 and doesn't have go-1.23.x
which is required now to build the binary. Switching to centos stream9
image make sure latest version of golang available to consume. Removing
curl because it is part of base image

In downstream we already have go-1.23 attached to our target tag which
make sure it build downstream also.

cfergeau commented on Containerfile.rpmbuild:
```diff
@@ -1,4 +1,4 @@
-FROM quay.io/almalinuxorg/8-base:8
+FROM quay.io/centos/centos:stream9
```
This can be problematic as we are not going to detect when we build against a libvirt version which is too new compared to what is in rhel8.
I'll reiterate my suggestion that we stop supporting rhel8 and switch to rhel9 minimum which would avoid this issue.

praveenkumar replied:
@cfergeau yes, should we wait till RHEL-10 release ?

cfergeau replied:
I'd be fine with https://libvirt.org/platforms.html
> The project aims to support the most recent major version at all times. Support for the previous major version will be dropped 2 years after the new major version is released or when the vendor itself drops support, whichever comes first. 

RHEL 9 was released in May 2024 almost 3 years ago, so we could follow the same policy and say we no longer support it. Especially as tools that we need (`oc`) have dropped support for it iirc.
But that's a team decision, not something to decide only here :)

praveenkumar committed:
libvirt: fix validatator error due to different value/variable type

This should fix following
```
go test ./...
pkg/libvirt/libvirt.go:71:13: cannot use memorySize (variable of type int) as uint value in assignment
pkg/libvirt/libvirt.go:91:10: cannot use int(cpus) (value of type int) as uint value in assignment
pkg/libvirt/libvirt.go:107:22: cannot use newDriver.Memory (variable of type uint) as int value in argument to d.setMemory
```

praveenkumar committed:
tools: Update golangci-lint to v1.64.7 version

praveenkumar committed:
CI: Update to go 1.23 in workflow

cfergeau commented on go.mod:
```diff
@@ -1,9 +1,9 @@
-	github.com/stretchr/testify v1.8.4
+	github.com/stretchr/testify v1.9.0
 	libvirt.org/go/libvirt v1.8000.0
 	libvirt.org/go/libvirtxml v1.9008.0
```
libvirtxml is native go code, so we can update it if we want, it's only `libvirt` which links to a particular libvirt version. Fine with me to keep this unchanged.

praveenkumar replied:
As of now let's keep this unchanged.

cfergeau commented on .github/workflows/compile.yml:
```diff
@@ -23,7 +23,7 @@ jobs:
         run: make CONTAINER_RUNTIME="docker" test-rpmbuild
       - name: Upload crc-driver-libvirt artifact
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
```
Would be good to also update the dependabot configuration so that it does this for us, see https://github.com/crc-org/vfkit/blob/aa3cf4c94b50834b5f569622e5daaded9e0c2ca3/.github/dependabot.yml#L13-L17 for an example of this.

cfergeau replied:
I see in one of the later commits that there's also a `tools` directory, we should also add it to dependabot config file.

praveenkumar replied:
Done, both are added.

cfergeau commented on Containerfile.stream9:
```diff
@@ -1,8 +1,8 @@
 MAINTAINER CRC team
 
-RUN microdnf install -y libvirt-devel curl git gcc make golang diffutils
+RUN dnf --enablerepo=crb install -y libvirt-devel curl git gcc make golang diffutils
```
Another option could be to keep using almalinux, and install go 1.23 ourselves from a tarball. Not sure it would be better though, just an alternative suggestion, feel free to ignore.

praveenkumar replied:
Ignored :)

cfergeau commented on Containerfile.stream9:
```diff
@@ -2,7 +2,7 @@ FROM quay.io/centos/centos:stream9
 MAINTAINER CRC team
 
-RUN dnf --enablerepo=crb install -y libvirt-devel curl git gcc make golang diffutils
+RUN dnf --enablerepo=crb install -y libvirt-devel git gcc make golang diffutils
```
This could be squashed in the previous commit

praveenkumar replied:
Done

cfergeau commented on pkg/libvirt/libvirt.go:
```diff
@@ -88,7 +88,7 @@ func (d *Driver) setVcpus(cpus uint) error {
 	}
 
-	d.CPU = int(cpus)
+	d.CPU = cpus
```
Most likely introduced by the update of `crc-org/machine`.

praveenkumar replied:
Yeah, right.

cfergeau commented on .github/workflows/compile.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
         uses: actions/setup-go@v3
         with:
-          go-version: 1.19
+          go-version: 1.23
```
imo this should be one of the very first commits in the PR as CI will be broken until then (not important, can be ignored, I suspect making the code "1.23" ready and then switching it to 1.23 for good, and ensuring everthing build at all steps will be more work than it is worth it)

praveenkumar replied:
Done

cfergeau reviewed (commented):
Apart from the comments asking for the dependabot config file to be extended, ack from me.

praveenkumar committed:
dependabot: Update config to put pr for gh action and tools/ directory

This will ensure updated version under `tools/go.mod` and also if some
github action updated.
